sklearn/utils/metadata_requests.py

```diff
@@ -0,0 +1,735 @@
+        ``{'alias': True}``. This is desired in meta-estimators passing
+        requests to their parent estimators.
+        """
+        res = MetadataRequest()
+        for method in METHODS:
+            setattr(res, method, getattr(self, method).masked())
+        return res
+
+    def to_dict(self):
+        """Return dictionary representation of this object."""
+        output = dict()
+        for method in METHODS:
+            output[method] = getattr(self, method).requests
+        return output
+
+    def __repr__(self):
+        return self.to_dict()
+
+    def __str__(self):
+        return str(self.to_dict())
+
+
+def metadata_request_factory(obj=None):
+    """Get a MetadataRequest instance from the given object.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    obj : object
+        If the object is already a MetadataRequest, return that.
+        If the object is an estimator, try to call `get_metadata_request` and get
+        an instance from that method.
+        If the object is a dict, create a MetadataRequest from that.
+
+    Returns
+    -------
+    metadata_requests : MetadataRequest
+        A ``MetadataRequest`` taken or created from the given object.
+    """
+    if obj is None:
+        return MetadataRequest()
+
+    if isinstance(obj, MetadataRequest):
+        return obj
+
+    if isinstance(obj, dict):
+        return MetadataRequest(obj)
+
+    try:
+        return obj.get_metadata_request(output="MetadataRequest")
+    except TypeError:
+        # The method exists, but doesn't accept `output`
+        return MetadataRequest(obj.get_metadata_request())
+    except AttributeError:
+        # The object doesn't have a `get_metadata_request` method.
+        return MetadataRequest()
+
+
+class MetadataRouter:
+    """Route the metadata to child objects.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init__(self):
+        self.requests = MetadataRequest()
+
+    def add(self, *obj, mapping="one-to-one", overwrite=False, mask=False):
+        """Add a set of requests to the existing ones.
+
+        Parameters
+        ----------
+        *obj : objects
+            A set of objects from which the requests are extracted. Passed as
+            arguments to this method.
+
+        mapping : dict or str, default="one-to-one"
+            The mapping between the ``obj``'s methods and this routing object's
+            methods. If ``"one-to-one"`` all methods' requests from ``obj`` are
+            merged into this instance's methods. If a dict, the mapping is of
+            the form ``{"destination_method": "source_method"}``.
+
+        overwrite : bool or str, default=False
+            - True: ``alias`` replaces the existing routing.
+            - False: a ``ValueError`` is raised if the given value conflicts
+              with an existing one.
+            - "on-default": only overwrite the alias if it is
+              RequestType.ERROR_IF_PASSED
+
+        mask : bool, default=False
```

Comment:
I'm trying to think of a better name for this.
* with_renames
* with_alias

not so happy yet.

I do get why it's needed now.